src/types.ts

```diff
@@ -2303,74 +2303,53 @@ export class ZodFunction<
       return INVALID;
     }
 
-    const handleArgs = (
-      args: any,
-      parsedArgs:
-        | { success: true; data: any }
-        | { success: false; error: ZodError }
-    ): any => {
-      if (!parsedArgs.success) {
-        const issue = makeIssue(
-          args,
-          pathToArray(ctx.path),
-          ctx.params.errorMap,
-          {
-            code: ZodIssueCode.invalid_arguments,
-            argumentsError: parsedArgs.error,
-          }
-        );
-        throw new ZodError([issue]);
-      }
-      return parsedArgs.data;
-    };
+    function makeArgsIssue(args: any, error: ZodError): ZodIssue {
+      return makeIssue(args, pathToArray(ctx.path), ctx.params.errorMap, {
+        code: ZodIssueCode.invalid_arguments,
+        argumentsError: error,
+      });
+    }
 
-    const handleReturns = (
-      returns: any,
-      parsedReturns:
-        | { success: true; data: any }
-        | { success: false; error: ZodError }
-    ): any => {
-      if (!parsedReturns.success) {
-        const issue = makeIssue(
-          returns,
-          pathToArray(ctx.path),
-          ctx.params.errorMap,
-          {
-            code: ZodIssueCode.invalid_return_type,
-            returnTypeError: parsedReturns.error,
-          }
-        );
-        throw new ZodError([issue]);
-      }
-      return parsedReturns.data;
-    };
+    function makeReturnsIssue(returns: any, error: ZodError): ZodIssue {
+      return makeIssue(returns, pathToArray(ctx.path), ctx.params.errorMap, {
+        code: ZodIssueCode.invalid_return_type,
+        returnTypeError: error,
+      });
+    }
 
     const params = { errorMap: ctx.params.errorMap };
     const fn = data;
 
     if (this._def.returns instanceof ZodPromise) {
       return OK(async (...args: any[]) => {
-        const parsedArgs = handleArgs(
-          args,
-          await this._def.args.spa(args, params)
-        );
+        const error = new ZodError([]);
```

Comment:
One thing I can share my experience on is not initializing an empty error. We did a lot of work to remove these allocations for performance reasons earlier in the year. Should be as simple as moving this into the catch blocks below.

Reply:
Very true though I think in this case it's equivalent to the current implementation. Both `handleArgs` and `handleReturns` create new ZodErrors currently.

Reply:
Hi @scotttrinh! I appreciate your response :). I'm not a pro at how node works internally, but strangely the stack trace is dramatically different depending on where the error is initialized. 

Stack Trace for Error Initialized in catch block 

(node:27924) UnhandledPromiseRejectionWarning: ZodError: [
  {
    "code": "invalid_return_type",
    "returnTypeError": {
      "issues": [
        {
          "code": "invalid_type",
          "expected": "string",
          "received": "number",
          "path": [],
          "message": "Expected string, received number"
        }
      ],
      "name": "ZodError"
    },
    "path": [],
    "message": "Invalid function return type"
  }
]
    at new ZodError (/Users/USERNAME/zod/src/ZodError.ts:134:5)
    at /Users/USERNAME/zod/src/types.ts:2335:27


Stack Trace for Error Initialized outside of catch block: 
```
(node:27945) UnhandledPromiseRejectionWarning: ZodError: [
  {
    "code": "invalid_return_type",
    "returnTypeError": {
      "issues": [
        {
          "code": "invalid_type",
          "expected": "string",
          "received": "number",
          "path": [],
          "message": "Expected string, received number"
        }
      ],
      "name": "ZodError"
    },
    "path": [],
    "message": "Invalid function return type"
  }
]
    at new ZodError (/Users/USERNAME/zod/src/ZodError.ts:134:5)
    at ZodFunction.<anonymous> (/Users/USERNAME/zod/src/types.ts:2325:23)
    at step (/Users/USERNAME/zod/src/types.ts:59:23)
    at Object.next (/Users/USERNAME/zod/src/types.ts:40:53)
    at /Users/USERNAME/zod/src/types.ts:34:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/USERNAME/zod/src/types.ts:30:12)
    at Object.numToString (/Users/USERNAME/zod/src/types.ts:1761:24)
    at callIt (/Users/USERNAME/zod/src/playground.ts:12:3)
    at callIt2 (/Users/USERNAME/zod/src/playground.ts:16:3)
```

Reply:
I wonder if this is related to returning a promise directly vs `return await`. https://www.even.li/posts/2020-06-04-async-stack-traces/ has some detail here.

Reply:
Eh, that didn't seem to make any difference. Going to poke at it some more.